### DISCUSSION TITLE
No configuration found for `rust`. No dap configs for rust

### DISCUSSION THREAD

**Discussion by BertrandD** [mrcjkb/rustaceanvim#236] [Q&A]
Hey ! 

I am using Lazy to load the plugin: 
```
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        rust_analyzer = function()
          return true 
        end,
      },
    },
  },
  {
    'mrcjkb/rustaceanvim',
    version = '^4', -- Recommended
    ft = { 'rust' },
    dependencies = {
      "mfussenegger/nvim-dap",
    },
    config = function ()
      local configs = require("plugins.configs.lspconfig")
      local on_attach = configs.on_attach
      
      vim.g.rustaceanvim = {
        dap = {
          autoload_configurations = true
        },
        server = {
          on_attach =  on_attach
        }
      }
    end
  },
```
  
When I run `require("dap").continue()` I get the error message 
```
No configuration found for `rust`. You need to add configs to `dap.configurations.rust`
```

checkhealth:
```

==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 0.3.1839-standalone
- OK Cargo: found cargo 1.70.0 (ec8a8a0ca 2023-04-25)
- OK rustc: found rustc 1.70.0 (90c541806 2023-05-31)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

I didn't found anything to unlock me on that...
I managed (sometimes) to have it working, but for unknown reason when I restart vim it stops working... And as I think I didn't changed anything in-between, I am confused...

Thank you for your help 🙏 , and for you amazing work !

**Comment by BertrandD** [comment]
Note: running `RustLsp runnables` and `RustLsp debuggables` works fine though, but would like to have it working with dap.continue if possible...

**Comment by mrcjkb** [comment]
Hey :wave: 

Just realised you already have `dap.autoload_configurations = true`, so I removed my previous comment :sweat_smile: 

rust-analyzer is a bit finicky, and can take a long time to finish initialising - especially for larger projects.
So when rustaceanvim sends a client request to autoload debug configurations, the server will respond with an empty list of configurations if it isn't fully initialised yet.
I use [`experimental/serverStatus`](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#server-status) to load DAP configurations when the language server is ready, in addition to `on_attach`.
But during startup, it can take quite some time before the server returns results.

I haven't fully understood why it's more reliable with `:RustLsp debuggables`. They both use exactly the same methods. I guess it's just that `:RustLsp debuggables` is less likely to be called before rust-analyzer is ready.

**Reply by BertrandD** [reply]
Hey :wave:

I have this issue with an simple hello world project, so I don't think it's the size of the project that causes the issue :confused: And even after some hours working using `:RustLsp debuggables` I still get the error

**Reply by BertrandD** [reply]
Well I just found something interesting:

I restarted my vim session and after a few seconds spamming `require("dap").continue()` it finally worked

BUT if I wait to get the first diagnostics *before* doing `require("dap").continue()` , then I can spamm as often as I want, I never get the configurations loaded ....

**Reply by BertrandD** [reply]
well, I didn't change anything... I went in a meeting, I come back and retry the same process described up there and now it works fine...

**Reply by mrcjkb** [reply]
rust-analyzer will start publishing diagnostics before it has fully initialized.
But it can take very long to initialize - it's [a known issue](https://github.com/rust-lang/rust-analyzer/issues/14258).

In some cases, ignoring some proc macros can help improve performance a bit:

```lua
vim.g.rustaceanvim = {
    -- ...
    server = {
      default_settings = {
        ['rust-analyzer'] = {
          procMacro = {
            enable = true, -- set this to false to disable completely
            ignored = {
              ['async-trait'] = { 'async_trait' },
              ['napi-derive'] = { 'napi' },
              ['async-recursion'] = { 'async_recursion' },
            },
          },
        },
      },
    },
  }
```
